src/lightning/fabric/fabric.py

```diff
@@ -575,19 +575,20 @@ def no_backward_sync(self, module: _FabricModule, enabled: bool = True) -> Gener
     def sharded_model(self) -> Generator:
         """Shard the parameters of the model instantly when instantiating the layers.
 
-        Use this context manager with strategies that support sharding the model parameters to save peak memory usage.
-
-        Example::
+        .. deprecated:: This context manager is deprecated in favor of :meth:`init_module`, use it instead.
+        """
+        rank_zero_deprecation("`Fabric.sharded_model()` is deprecated in favor of `Fabric.init_module()`.")
+        with _old_sharded_model_context(self._strategy):
+            yield
 
-            with self.sharded_model():
-                model = MyModel()
+    @contextmanager
+    def init_module(self) -> Generator:
+        """Instantiate the model and its parameters under this context manager to reduce peak memory usage.
```

Comment:
This might also be used in the future for things that arent modules, maybe we should choose a different name that doesnt explicitly state "module"

Reply:
I was thinking mentioning the "module" in the name is safer because it expresses more precisely what we should do under this context manager. This hopefully helps prevent users from accidentally doing weird stuff under this context manager and getting in trouble, but of course, we can never completely prevent that.

Reply:
I would vote for something like `efficient_init` or similar

Reply:
For example, this context manager woudl be useful for non-modules too:

```python
import torch
from lightning import Fabric

fabric = Fabric(accelerator="cuda", precision="64-true")
with fabric.init_module():
    x = torch.zeros(1)
    print(x.device)  # cuda
    print(x.dtype)  # 64

```

Reply:
It is very important that this context manager be only used for model initialization. For everything else, the user should use `fabric.device`.

The main motivation for this is to load and shard large models efficiently, or to provide a convenient way to cast the model to the desired dtype for inference without code changes.